luci-app-wol: persist the fields values

mdevolde opened the pull request:
- [x] This PR is not from my *main* or *master* branch :poop:, but a *separate* branch :white_check_mark:
- [x] Each commit has a valid :black_nib: `Signed-off-by: <my@email.address>` row (via `git commit --signoff`)
- [x] Each commit and PR title has a valid :memo: `<package name>: title` first line subject for packages
- [x] Incremented :up: any `PKG_VERSION` in the Makefile
- [x] Tested on: aarch64, OpenWrt 24.10.4, Opera One(version : 125.0.5729.21) :white_check_mark:
- [x] Mention: @mdevolde 
- [x] Screenshot of changes:

#### Display when loading the app:
<img width="1277" height="560" alt="image" src="https://github.com/user-attachments/assets/9f7fd10b-f85b-411f-a8ae-79bdd686eddf" />

#### Modification of interface and broadcast fields:
<img width="1271" height="570" alt="image" src="https://github.com/user-attachments/assets/1fc05066-d562-49e3-9d30-51f3e10fd955" />

#### Wake up the host by pressing the button:
<img width="1228" height="534" alt="image" src="https://github.com/user-attachments/assets/92fcb017-5424-4dfd-9702-43278bd64210" />

#### Ability to persist the new values of modified fields:
<img width="1209" height="544" alt="image" src="https://github.com/user-attachments/assets/9f923611-4549-44bf-a135-cf107b35c32b" />


- [x] Closes: #8145 
- [x] Description:

This PR adds optional persistence of field values to the `luci-app-wol` application.

Persistence is implemented using a dedicated UCI configuration.
After a device wake-up, if the user has modified any fields, these values can be saved and reused. When persistence is enabled, the previously saved values are automatically preselected the next time the application page is opened.

In addition, this PR adds a runtime check for the presence of at least one supported wake-up utility (`etherwake` or `wakeonlan`). If neither binary is available, a warning message is displayed to the user.

Regarding ACLs:
- `"uci": [ "etherwake" ]` in read mode has been removed, as it was used to find a default interface
- `"uci": [ "luci-wol" ]` in read mode has been added to allow the new dedicated configuration to be read
- `"uci": [ "luci-wol" ]` and `"ubus": { "uci": [ "add", "set", "delete" ] }`, have been added in write mode to allow modification of this new dedicated configuration

Signed-off-by: Martin Devolder [martin.devolder2@gmail.com](mailto:martin.devolder2@gmail.com)

systemcrash commented on applications/luci-app-wol/htdocs/luci-static/resources/view/wol.js:
```diff
@@ -12,157 +10,253 @@ const ETHERWAKE_BIN = '/usr/bin/etherwake';
+                L.toArray(hosts[mac].ip6addrs || hosts[mac].ipv6)[0] ||
+                '?'
+            ]), ')']));
+        });
```
This is how it's done in `dhcp.js`
```js
		Object.keys(hosts).forEach(function(mac) {
			for (const ip of L.toArray(hosts[mac].ipaddrs || hosts[mac].ipv4))
				ipaddrs[ip] = hosts[mac].name || mac;
		});
		L.sortedKeys(ipaddrs, null, 'addr').forEach(function(ipv4) {
			o.value(ipv4, ipaddrs[ipv4] ? '%s (%s)'.format(ipv4, ipaddrs[ipv4]) : ipv4);
		});
```

mdevolde replied:
@systemcrash I tried to clean up my approach by drawing inspiration from this, but not by simply copying it, because the situation is not exactly the same.

mdevolde commented on applications/luci-app-wol/htdocs/luci-static/resources/view/wol.js:
```diff
@@ -12,157 +10,253 @@ const ETHERWAKE_BIN = '/usr/bin/etherwake';
+            bin = ETHERWAKE_BIN;
+
+        // Build argument list based on selected binary
+        var args = [];
```
done

mdevolde commented:
@systemcrash your revisions have been applied.

systemcrash commented on applications/luci-app-wol/htdocs/luci-static/resources/view/wol.js:
```diff
@@ -12,157 +10,232 @@ const ETHERWAKE_BIN = '/usr/bin/etherwake';
+        if (has_ewk) {
+            o = s.option(form.Flag, 'broadcast', _('Broadcast'), _('Etherwake only'));
+            o.default = '0';
+        }
```
Add a flag for debug

mdevolde replied:
I find that the result is much more meaningful for the end user in debug mode. What are the disadvantages of doing this? I nevertheless did add the return code retrieval with `close` in the ucode.

systemcrash commented on applications/luci-app-wol/htdocs/luci-static/resources/view/wol.js:
```diff
@@ -12,157 +10,232 @@ const ETHERWAKE_BIN = '/usr/bin/etherwake';
+        const args = [];
+
+        if (bin == ETHERWAKE_BIN) {
+            args.push('-D');
```
set debug flag based on section_id pref

mdevolde replied:
Same as https://github.com/openwrt/luci/pull/8147#discussion_r2617377700

systemcrash commented:
If you use a TableSection, you can have everything available to edit in the overview:
<img width="1216" height="224" alt="Screenshot 2025-12-14 at 19 09 09" src="https://github.com/user-attachments/assets/a9035a89-1586-4378-8f5f-122b135756da" />
with `s.nodescriptions = true;` to hide the description row.

systemcrash commented on applications/luci-app-wol/htdocs/luci-static/resources/view/wol.js:
```diff
@@ -12,157 +10,232 @@ const ETHERWAKE_BIN = '/usr/bin/etherwake';
+                'class': 'alert-message warning'
+            }, [
+                E('p', _('No Wake on LAN utilities found. Please install either the "etherwake" or "wakeonlan" package.'))
+            ]);
```
There you can skip the fs.exec_direct calls and simply rely on the `has_ewk` values already.

mdevolde replied:
done. I have also added the "redirection button" in case only one of the two binaries is installed, to make it easy to find the other one.

mdevolde commented on applications/luci-app-wol/htdocs/luci-static/resources/view/wol.js:
```diff
@@ -12,157 +10,232 @@ const ETHERWAKE_BIN = '/usr/bin/etherwake';
+    }),
+
+    callHostHints: rpc.declare({
+        object: 'luci-rpc',
```
done

mdevolde commented on applications/luci-app-wol/htdocs/luci-static/resources/view/wol.js:
```diff
@@ -12,157 +10,232 @@ const ETHERWAKE_BIN = '/usr/bin/etherwake';
+                    E('strong', {}, has_ewk ? 'Etherwake' : 'Wakeonlan')
+                ]);
+            };
+            o.rawhtml = true;
```
done

mdevolde commented:
@systemcrash I added `nodescriptions`, but isn't it overkill to use `TableSection` since everything can be edited with the edit button?

Your reviews have been applied, except for the one concerning debug mode (I have added a comment above).

mdevolde commented:
@systemcrash your last comment is applied.
